### DISCUSSION TITLE
How to return the object during UPSERT? Also, why "statusText" is "Created" even during update?

### DISCUSSION THREAD

**Discussion by DNA-PC** [PostgREST/postgrest#2760] [Q&A] [answered]
At https://postgrest.org/en/stable/api.html#upsert I noticed two things:
- Response always returns `statusText = "Created"`, even on update. How do I know if it was an insert/update then?
- Response doesn't contain the updated object, is there an easy way to fetch it, or must it be done in another request?

**Comment by steve-chavez** [comment, marked as answer]
That seems related to https://github.com/supabase/postgrest-js/ and not PostgREST itself. Ask on https://github.com/orgs/supabase/discussions/.

Our error format is documented at https://postgrest.org/en/latest/errors.html?highlight=504#error-source

**Reply by Vadorequest** [reply]
_Edit: (I'm the same user, different account)_

I really don't know what's being used underneath.
Other people pointed me to PostgREST when calling the REST API endpoints directly, and I call the API endpoints directly without going through any lib (vanilal JS)

Here are my logs
![image](https://user-images.githubusercontent.com/3807458/233421819-37e79311-6cfb-4705-a784-a5f4fd8fefb5.png)


Using this code:

```js
  console.log(`Fetching "${EMPLOYEE_UPSERT_ENDPOINT_URL}"`);
  const response = await fetch(`${EMPLOYEE_UPSERT_ENDPOINT_URL}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'apikey': SUPABASE_ANON_KEY,
      'Prefer': 'resolution=merge-duplicates',
    },
    body: JSON.stringify(employeeProperties)
  });

  // Log the response for debugging purposes.
  console.log(`response`, response);
```

**Reply by steve-chavez** [reply]
> Response doesn't contain the updated object, is there an easy way to fetch it, or must it be done in another request?

Use `Prefer: return=representation`. https://postgrest.org/en/stable/api.html#insertions

> Response always returns statusText = "Created", even on update. How do I know if it was an insert/update then?

Check https://github.com/PostgREST/postgrest/issues/1683, `xmax` might help.

**Reply by DNA-PC** [reply]
Thank you.

`'Prefer': 'resolution=merge-duplicates, return=representation',` did the trick.

Regarding `xmax`, it's a bit fuzzy, I did it like this:

```js
const EMPLOYEE_UPSERT_ENDPOINT_URL = `${SUPABASE_URL}/rest/v1/${TABLE_NAME}?on_conflict=airtableId&select=id,airtableId,xmax`; // Merge duplicates using their airtableId

...

const responseJSON = await response.json();
    const upsertedRecord = responseJSON?.[0];

    if (upsertedRecord?.xmax !== '0') {
      console.log('Employee record updated successfully', upsertedRecord);
    } else {
      console.log('Employee record created successfully', upsertedRecord);
    }

```

It works. 

I wish I could return "xmax" without specifying it in `select` (querystring).
I tried `select=*,xmax` but got `{code: "42703", details: null, hint: null, message: "column Employee.xmax does not exist"}`

I can live with that, though.
